16638 & 16639 Created new BulletsColinLink component + implemented new registration wireframes

JazzarKarim opened the pull request:
*Issue #:* bcgov/entity#16638 & bcgov/entity#16639

*Description of changes:*
- Created the new BulletsColinLink component
- Display component when a business type is selected
- Created new enum `CompanyType`
- New radio button group
- Named Company --> Old existing flow
- Numbered Company --> New flow
- Two buttons: "Incorporate Now" and "Go to Corporate Online to Register"
- Display button according to the FF `supported-incorporation-registration-entities`
- Fixed styling and look
- Incorporate Now button creates a draft and redirects to Entities Dashboard
- Go to Corporate Online to Register button redirects to Colin
- New component has 3 props and 1 emitter. Please see comments for definition
- app version: 5.0.6

*Update:*
- Created new dialog: `IncorporateNowErrorDialog`. Show this dialog when user incorporate now fails after user signs in. This also handles the error
- Added Unit Tests
- Utilized the `CorpTypeCd` enum.

Note: These changes implement the new registration wireframes in the UXPin (https://preview.uxpin.com/a86dccdd20ac62828f965c5eea10f81b828a9853#/pages/163136223/simulate/sitemap)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the namerequest license (Apache 2.0).

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

JazzarKarim commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+// Font weight + size of button. This is done to stay consistent with "Check this Name" Button.
+.v-btn {
+  font-weight: bold;
+  min-height: 45px;
```
This value is the same value for the button in `search.vue`. I thought it would be a good idea to keep values consistent.

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

seeker25 commented:
Lookin good Karim

JazzarKarim commented:
> Lookin good Karim

Thanks Travis!!

severinbeauvais commented:
Do you have guidance from UX to capitalize Name and Designation? If not then please follow the uxpins.

![image](https://github.com/bcgov/namerequest/assets/10002889/e1e8dc73-edf9-424e-a0dd-d88578f89830)

severinbeauvais commented:
One time I tried to incorporate now, the API gave a 503 error, but no error was displayed in Namerequest UI. Please check that errors are correctly caught.

![image](https://github.com/bcgov/namerequest/assets/10002889/2b3d5a34-f819-49bb-b634-1f40370a40ee)

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+<template>
+  <div id="bullets-colin-link">
```
You probably don't need this div, and can use v-row as the main element in this template.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+  <div id="bullets-colin-link">
+    <v-row>
+      <v-col cols="12" sm="3">
+        <v-radio-group v-model="selectedCompanyType" @change="radioButtonChanged" flat mandatory>
```
I like to see event handlers with `()` to emphasize that they're functions.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+import { CompanyType } from '@/enums'
+import NameInput from '@/components/new-request/name-input.vue'
+import { Navigate } from '@/plugins'
+import BusinessServices from '@/services/business.services'
```
Can we rename the `xxx.services` files to `xxx-services` (and keep the period for the extension only)?

JazzarKarim replied:
I see you've done this change in your PR Sev. I'll not be making this change in this PR then OK?

severinbeauvais replied:
Yup.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+})
+export default class BulletsColinLink extends Vue {
+  /** The selected business type. */
+  @Prop({ default: '' }) readonly businessType!: string
```
Does this have an enum for type?

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+    if (entityType === 'CR') return 'BEN'
+    if (entityType === 'UL') return 'ULC'
+    return ''
+  }
```
These are not names -- they're codes.

I'd like to see enums used, please.

severinbeauvais replied:
Great comment. Thanks!

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+  // For template
+  readonly CompanyType = CompanyType
+
+  // Get radio text respective to Enum values of CompanyType.
```
If you use JSDoc format then the comment here will appear in a tooltip when you hover over the method.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+    } else {
+      const registryHomeUrl = sessionStorage.getItem('REGISTRY_HOME_URL')
+      const nameRequestUrl = `${window.location.origin}`
+      Navigate(`${registryHomeUrl}login?return=${nameRequestUrl}`)
```
I believe you have to put the current data in the store so it's restored when you return to Namerequest UI. This is done elsewhere in this project -- search for it.

severinbeauvais replied:
See existing-request-display.vue:705.

severinbeauvais replied:
Looks good, but why is it `Navigate()` here but `navigate()` in existing-request-display.vue?

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+        this.goToEntityDashboard(businessId)
+      } catch (error) {
+        this.$root.$emit('showSpinner', false)
+        throw error
```
You should show the error here. Throwing it, with nobody to catch it, will display something in the console, but the user won't otherwise know that something went wrong.

severinbeauvais replied:
I think this got fixed when you moved `incorporateNow()` to nr-affiliation-mixin.ts.

severinbeauvais replied:
Actually, I'm not sure. Does error handling work now?

JazzarKarim replied:
No Sev, it does not. I still need to fix it. Throwing a new Error is still showing it in the console. I may have to use the Error dialog.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+    if (entityType === 'BC' || entityType === 'CC') return entityType
+    if (entityType === 'CR') return 'BEN'
+    if (entityType === 'UL') return 'ULC'
+    return ''
```
I think this would be cleaner and more extensible as a switch-case statement.

If the entity type does not need to be overridden then I would just return it (instead of specifically handling BC and CC or returning empty string).

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+   */
+  async incorporateNow () {
+    if (this.getIsAuthenticated) {
+      try {
```
Canyou extend the existing nr-affiliation-mixin to handle numbered IAs, instead of duplicating code here?

severinbeauvais replied:
You've moved this to the mixin. I assume you tried to reuse code and it wasn't coming together, so that's fine. Resolving.

severinbeauvais commented on src/components/new-request/search.vue:
```diff
@@ -212,7 +182,43 @@
+      @radioButtonChange="selectedCompanyType = $event"
+    >
+      <template v-slot:name-input-slot>
+        <NameInput
```
I know this code is the same as before, but please call it out for verification in your ticket. Thx.

severinbeauvais replied:
Karim, you can resolve this when you've done it.

severinbeauvais commented on src/components/new-request/search.vue:
```diff
@@ -311,8 +314,10 @@ export default class NewSearch extends Mixins(CommonMixin) {
   corpNumValid = true
-  readonly corpOnlineLink = 'https://www.corporateonline.gov.bc.ca/'
   locationDisabled = false
+  selectedCompanyType = ''
```
This declares the type as string. It should be CompanyType instead. (You'll probably need to init as null.)

severinbeauvais reviewed (commented):
Looks really good (and works, too)! 😃 

Is there any chance you could create 2 unit tests: one to verify the display of the search page for a named company (showing Check this Name button), and for a numbered company (showing Incorporate Now button)?

JazzarKarim commented:
> Looks really good (and works, too)! 😃
> 
> Is there any chance you could create 2 unit tests: one to verify the display of the search page for a named company (showing Check this Name button), and for a numbered company (showing Incorporate Now button)?

Thanks Sev!

I have fixed most of the comments but there are still a few left. I am still getting the 503 error due an Axios Error sometimes. I need to fix that in addition to some styling and unit tests.

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,165 @@
+      case EntityType.BC: return 'BEN'
+      case EntityType.CR: return 'BC'
+      case EntityType.UL: return 'ULC'
+      default: return entityType
```
Since CorpTypeCd is now imported in this app, you should use that enum instead of strings.

You can also use this type in incorporateNow() and createBusinessIA().

JazzarKarim replied:
Please let me know if we can resolve this conversation Sev. Ditto the one below it.

severinbeauvais replied:
Yup, code looks good. Resolving.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,212 @@
+
+<style lang="scss" scoped>
+@import '@/assets/styles/theme.scss';
+// Radio button group spacing
```
Karim, you can resolve this.

severinbeauvais commented on src/components/new-request/search.vue:
```diff
@@ -212,7 +182,43 @@
+          register your extraprovincial business immediately using its existing name at Corporate Online.</p>
+      </template>
+      <template v-slot:designation>
+        <v-col md="4" lg="4">
```
You probably don't need the `lg` -- how this works is columns are 12 wide (unless you declare them differently) and are 4 wide from `md` size and up.

severinbeauvais commented on src/App.vue:
```diff
@@ -214,6 +214,14 @@ export default class App extends Mixins(
+      // clear the legal type data
+      sessionStorage.removeItem('LEGAL_TYPE')
+    }
+
```
Check also Signin.vue as it may need similar code.

JazzarKarim replied:
I tested this extensively Sev. Apparently, the code in Signin.vue is not being executed when I go with this flow. So, I only added the code to App.vue. Not sure if I'm missing a case in which the code in Signin.vue would run.

severinbeauvais replied:
What scenarios did you test?

Did you put a breakpoint (or console.log) in Signin.vue to see when it's used?

It's possible that code in Signin.vue is a "decoy" (looks like it's doing something but really isn't) but I'm not sure.

severinbeauvais replied:
Here is what causes Signin.vue to be rendered:
https://github.com/bcgov/namerequest/blob/92172b3bf0756722475bf804249b0979ed624664/src/router/routes.ts#L25

severinbeauvais replied:
Resolved.

JazzarKarim commented:
## Latest commit:
- Created new dialog: `IncorporateNowErrorDialog`. Show this dialog when user incorporate now fails after user signs in. This also handles the error
- Fixed styling and sizing
- Added Unit Tests

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

JazzarKarim commented:
> Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

For some reason, I can't see the feature flags in the temporary URL.

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

severinbeauvais commented:
> * Created new dialog: `IncorporateNowErrorDialog`. Show this dialog when user incorporate now fails after user signs in. This also handles the error

I think there is only 1 error dialog shown app-wide, so here you are establishing a new pattern. Check with the UX team on this. (Originally there were more and more detailed errors displayed but the UX team wanted them simplified. I'm not sure giving detail here is appropriate.)

On the other hand, if your dialog is generic enough, could it be reused for the "firm register now" code?

> * Added decoy code to `Signin.vue`

"Decoy pattern" is not something you want to do, as it makes life harder for future devs. I'd rather see unused code deleted :)  Were you able to confirm whether the code in Signin is ever used?

JazzarKarim commented:
> > * Created new dialog: `IncorporateNowErrorDialog`. Show this dialog when user incorporate now fails after user signs in. This also handles the error
> 
> I think there is only 1 error dialog shown app-wide, so here you are establishing a new pattern. Check with the UX team on this. (Originally there were more and more detailed errors displayed but the UX team wanted them simplified. I'm not sure giving detail here is appropriate.)
> 
> On the other hand, if your dialog is generic enough, could it be reused for the "firm register now" code?
> 
> > * Added decoy code to `Signin.vue`
> 
> "Decoy pattern" is not something you want to do, as it makes life harder for future devs. I'd rather see unused code deleted :) Were you able to confirm whether the code in Signin is ever used?

- I will check with the UX team if it's OK. I was thinking of the best way to handle the errors and this seemed the nicest way. I'll confirm with the UX team however. Also, yes I believe the dialog should be reusable for the register now code for firms.
- Regarding the decoy pattern, I don't see the code for Signin.vue being used in the context of this PR. I tried debugging and testing by logging in and out many times. The code there is not being executed. Shall I, in this case, remove the unused code there?

JazzarKarim commented:
> * I will check with the UX team if it's OK. I was thinking of the best way to handle the errors and this seemed the nicest way. I'll confirm with the UX team however. Also, yes I believe the dialog should be reusable for the register now code for firms.

Just confirmed with the Janis. The dialog is OK. She said it's fine. However, I'm going to make a little adjustment to it.

severinbeauvais commented:
> I believe Signin.vue is used by the login button in the header. Please try that. I'll try to confirm as well.

Confirmed. If you are in Namerequest UI, not logged in, and you login, then Signin.vue is used and onReady() is called.

However, it seems weird to me that we would need to check whether to restore a NR in session storage here. That's done in App.vue, isn't it? Or maybe it is done here and the code in App.vue isn't used? You'll need to test various scenarios to see which code is used and which isn't. I'm pretty sure we don't need both.

JazzarKarim commented:
> If you're certain it's not used then, yes, unused (decoy) code should be deleted.
> 
> I believe Signin.vue is used by the login button in the header. Please try that. I'll try to confirm as well.

Confirmed, you're right Sev 😄 Apologies for that, I missed it. I'll be leaving the code in `Signin.vue`.

JazzarKarim commented:
> > If you're certain it's not used then, yes, unused (decoy) code should be deleted.
> > I believe Signin.vue is used by the login button in the header. Please try that. I'll try to confirm as well.
> 
> Confirmed, you're right Sev 😄 Apologies for that, I missed it. I'll be leaving the code in `Signin.vue`.

However, I don't see a scenario where we need to restore the legal type from session storage after signing in from the header. That's why I'm leaning more to removing it.

JazzarKarim commented:
> However, it seems weird to me that we would need to check whether to restore a NR in session storage here. That's done in App.vue, isn't it? Or maybe it is done here and the code in App.vue isn't used? You'll need to test various scenarios to see which code is used and which isn't. I'm pretty sure we don't need both.

Yes, I totally agree.

severinbeauvais commented:
> However, I don't see a scenario where we need to restore the legal type from session storage after signing in from the header. That's why I'm leaning more to removing it.

Exactly. I believe the only time we put that data in session storage is to programmatically get them to login, and when we return to the app then App.vue continues the affiliation/draft. Conversely, if the user decides to login at some point, there shouldn't be that data in session storage.

Anyways, you got this. Clean up if possible. Thanks.

severinbeauvais commented on src/components/dialogs/incorporate-now-error.vue:
```diff
@@ -33,6 +34,11 @@ export default class IncorporateNowErrorDialog extends Vue {
+
+  /** Try again button clicked. Refresh the page. */
+  tryAgain () {
+    window.location.reload()
```
This seems harsh. Does it work or did the session storage data get nuked already?

What do other error modals do?

JazzarKarim replied:
I've tested and it works! At this point in time, the `LEGAL_TYPE` session storage item is still there. So, upon a refresh, `App.vue` sees the item and moves forward with the regular process.

JazzarKarim commented:
> Exactly. I believe the only time we put that data in session storage is to programmatically get them to login, and when we return to the app then App.vue continues the affiliation/draft. Conversely, if the user decides to login at some point, there shouldn't be that data in session storage.

My thoughts exactly. All right, sounds good Sev!

severinbeauvais commented on src/App.vue:
```diff
@@ -61,6 +61,11 @@
+    <IncorporateNowErrorDialog
+      attach="#app"
+      :dialog="incorporateNowErrorDialog"
+      @close="incorporateNowErrorDialog = false"
```
What state does Close put the user in -- what can they do after this error?

JazzarKarim replied:
Back to square zero. They'll be going back to the main page and can try again from scratch.

Actually, now that you've mentioned this Sev, I might have to clear the session storage variable when the user closes the dialog. I'll be adding that.

severinbeauvais commented:
> It would be nice if the designation went to the next line on a narrow screen:

I see you've fixed this 👍 

![image](https://github.com/bcgov/namerequest/assets/10002889/0362015a-73ee-472c-985b-3b7ba80fe8bf)

JazzarKarim commented:
/gcbrun

severinbeauvais commented on src/components/dialogs/incorporate-now-error.vue:
```diff
@@ -32,8 +32,11 @@ export default class IncorporateNowErrorDialog extends Vue {
+  /** Pass click event to parent. Remove the session storage LEGAL_TYPE variable. */
+  @Emit()
+  close () {
+    sessionStorage.removeItem('LEGAL_TYPE')
```
To keep this dialog naive, I think the parent should clear the session storage.

severinbeauvais replied:
-> "Simple components" that should not know the big picture or business logic. This component should just be an error dialog, with the work done elsewhere.

JazzarKarim replied:
Makes a lot of sense. On it Sev.

severinbeauvais replied:
Looks good.

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

JazzarKarim commented:
@severinbeauvais In my latest commit, I moved the error status of incorporate now to the store. This is linked to the error dialog.

Reason: I want to show the error dialog whenever I want. This provides much more flexibility when we want to show the dialog. I suspect this is going to be helpful also when working on the other wireframes. Is that OK with you?

severinbeauvais commented:
> @severinbeauvais In my latest commit, I moved the error status of incorporate now to the store. This is linked to the error dialog.
> 
> **Reason:** I want to show the error dialog whenever I want. This provides much more flexibility when we want to show the dialog. I suspect this is going to be helpful also when working on the other wireframes. Is that OK with you?

OK with me! I appreciate the thought you put into this. Looks good.

severinbeauvais commented on src/components/common/bullets-colin-link.vue:
```diff
@@ -0,0 +1,175 @@
+  /** Emit the selected radio button CompanyType enum value. */
+  @Emit('radioButtonChange')
+  radioButtonChanged (event: CompanyType): CompanyType {
+    return event
```
FYI, there's an alternate syntax that does the same thing:
```
@Emit('radioButtonChange')
radioButtonChanged (event: CompanyType): void {}
```

JazzarKarim replied:
I like this 😄 I changed it!

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

JazzarKarim commented:
@severinbeauvais Please let me know if there are any outstanding comments Sev. Sorry about that, I know this work has been going on for a while.

severinbeauvais reviewed (approved):
LGTM. Would be nice to get another approval or 2 :)

JazzarKarim commented:
/gcbrun

pwei1018 commented:
Temporary Url for review: https://namerequest-dev--pr-620-gvxwe8bm.web.app

leodube-aot reviewed (approved):
LGTM Karim

chenhongjing reviewed (approved):
Lots of work! LGTM 😸
